### ISSUE TITLE
MiniMax H3: shared cache_directory leaks validation samples into training buckets

### ISSUE THREAD

**Issue by lcdrzxlcdrzx** [kohya-ss/musubi-tuner#1060]
### Summary

When a training dataset and a validation dataset use separate JSONL files but the same `cache_directory`, MiniMax H3 training does not honor the JSONL split at the bucket-building stage. It globs every `*_mmh3.safetensors` file in the shared cache directory, causing validation samples to be included in the training dataloader.

### Environment

- MiniMax H3 Ref2VA-Omni LoRA training
- 64 cached videos, split deterministically into 56 training and 8 validation items
- separate `train_56.jsonl` and `val_8.jsonl`
- initially, both TOML files pointed to the same cache directory

### Reproduction

1. Build 64 latent/text cache pairs in one directory.
2. Create two JSONL files: one with 56 video paths and one with the remaining 8 paths.
3. Configure the training TOML with the 56-item JSONL and the shared cache directory.
4. Start training.
5. Log output reports `loaded 56 videos` from the JSONL but then bucket counts and `num train items` equal 64, not 56.

### Root cause

In `src/musubi_tuner/dataset/image_video_dataset.py`, bucket construction enumerates the cache directory with a glob similar to:

```python
latent_cache_files = glob.glob(os.path.join(self.cache_directory, f"*_{self.architecture}.safetensors"))
```

It then builds ItemInfo from every matched cache file. It does not restrict the cache files to sample identifiers supplied by the configured JSONL datasource.

### Verified workaround

Use independent cache directories for train and validation. We moved cache pairs by video stem into:

- train cache: 56 latent + 56 text files
- validation cache: 8 latent + 8 text files

After updating each TOML to its matching cache directory, the log correctly reported:

```text
loaded 56 videos
total batches: 56
num train items: 56
```

### Proposed fix

Construct bucket inputs from the datasource's resolved item list, or filter discovered cache files to the item keys present in the configured JSONL. A trainer should never silently include unrelated cache files just because they exist in the cache directory.

### Why this matters

The current behavior invalidates validation loss: validation samples leak into training even though users supplied separate JSONL files. This is especially harmful for small video datasets, where validation loss is a primary signal for overfitting.

---

**Comment by kohya-ss** [comment]
Thank you for the report. This is not specific to MiniMax-H3: every architecture builds its training items by globbing `cache_directory` for `*_{arch}.safetensors`, so a cache directory shared between two dataset configs (train / validation in your case) mixes their items. The documented assumption is one cache directory per dataset; within a single config the loader already rejects duplicates.

#1093 (by @rossnot) restricts the glob to the items declared by the JSONL datasource, which fixes exactly this case. I'll track the fix there and close this issue as a duplicate once it is merged. Until then, separate cache directories (your verified workaround) are the recommended setup.
